### PR TITLE
feature: Add UbqScaledImageGenerator

### PR DESCRIPTION
**Pull request by maltebaumann**
Adds `UbqScaledImageGenerator` which provides the following:
- Given an initial `maxImageSize` and an encoded image, this generator will only yield downscaled versions of the contained image, that have at most an edge length of `maxImageSize`
- It attempts to use dimensions, that can be directly decoded from the encoded image representation, which may not fully fill the size defined by `maxImageSize`, but gives a performance benefit
- If the codec doesn't support decoding into matching dimensions, falls back to in-memory downscaling via `scalePixels`, but discards the full copy of the decoded image (-> tested by disabling codec-based dimension lookup)
- Accounts for EXIF orientation (-> tested with our beloved EXIF test images)

### CODE REVIEW

File: src/core/UbqScaledImageGenerator.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    // Alter the advertised dimensions to fit into maxImageSize
+    if (info.width() > maxImageSize || info.height() > maxImageSize) {
+        float width = info.width();
+        float height = info.height();
+        float maxSize = maxImageSize;
+        const auto desiredScale = std::min(maxSize / width, maxSize / height);
+
+        // Query codec for supported dimensions
+        const auto supportedDimensions = codec->getScaledDimensions(desiredScale);
```
Comment (keiwando) [inline src/core/UbqScaledImageGenerator.cpp:34]: How much smaller than the desired scale might this be? Do we need to add a lower limit here?

**Review by keiwando — APPROVED** [review]

File: src/core/UbqScaledImageGenerator.cpp (hunk 1 above)
Comment (maltebaumann) [reply src/core/UbqScaledImageGenerator.cpp:34]: Good point, I couldn't find a definitive answer. Could add a check that ensures that at least one edge has 75% of `maxSize`.

File: src/core/UbqScaledImageGenerator.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                    return image->scalePixels(pm, SkSamplingOptions(SkFilterMode::kLinear));
+                }
+
+                return false;
+            }
+        };
+    }
+
+    // Account for EXIF
+    if (decode) {
```
Comment (thierryseegers) [inline src/core/UbqScaledImageGenerator.cpp]: It appears to me that `decode` will always be set.

File: src/core/UbqScaledImageGenerator.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2015 Google Inc.
```
Comment (thierryseegers) [inline src/core/UbqScaledImageGenerator.cpp]: I'm not a lawyer but I would've expected to see our own copyright notice here.

File: src/core/UbqScaledImageGenerator.cpp (hunk 1 above)
Comment (maltebaumann) [reply src/core/UbqScaledImageGenerator.cpp:34]: Pushed a fix, that ensures at least 75% of `maxImageSize` is met.

**Review by thierryseegers — APPROVED** [review]

File: src/core/UbqScaledImageGenerator.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    if (!needsScaling()) {
+        return fCodec->queryYUVAInfo(supportedDataTypes, yuvaPixmapInfo);
+    }
+
+    return false;
+}
+
+bool UbqScaledImageGenerator::onGetYUVAPlanes(const SkYUVAPixmaps& yuvaPixmaps) {
+    if (!needsScaling()) {
+        switch (fCodec->getYUVAPlanes(yuvaPixmaps)) {
```
Comment (Bigpet) [inline src/core/UbqScaledImageGenerator.cpp:144]: not that it matters, but I think you can use `return isValidDecode(fCodec->getYUVAPlanes(yuvaPixmaps));` here as well

File: src/core/UbqScaledImageGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    } else {  // Query scaled version
+        decode = [this, options, info](const SkPixmap& pm) {
+            // Ask codec to return image at requested size
+            SkCodec::Result getPixelsResult = fCodec->getPixels(pm, options);
+            if (isValidDecode(getPixelsResult)) {
+                return true;
+            } else {
+                // Fall back to decode at closest dimension followed by forced downscale
+                float limit = fMaxImageSize;
+                float width = fCodec->getInfo().width();
```
Comment (Bigpet) [inline src/core/UbqScaledImageGenerator.cpp:103]: I know it's paranoid, but I always fear the divide by 0
```suggestion
                float width = std::max(1,  fCodec->getInfo().width());
```

File: src/core/UbqScaledImageGenerator.cpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                float height = fCodec->getInfo().height();
```
Comment (Bigpet) [inline src/core/UbqScaledImageGenerator.cpp:104]: same here 
```suggestion
                float height = std::max(1, fCodec->getInfo().height());
```

File: include/core/UbqScaledImageGenerator.h [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+#include "include/core/SkData.h"
+#include "include/core/SkImageGenerator.h"
+
+/**
+ * Given an initial `maxImageSize` and an encoded image, this generator will only yield downscaled
+ * versions of the contained image, that have at most an edge length of `maxImageSize`
+ * It attempts to use dimensions, that can be directly decoded from the encoded image
+ * representation, which may not fully fill the size defined by `maxImageSize`, but gives a
+ * performance benefit If the codec doesn't support decoding into matching dimensions, falls back to
+ * in-memory downscaling via `scalePixels`, but discards the full copy of the decoded image Accounts
```
Comment (Bigpet) [inline include/core/UbqScaledImageGenerator.h:14]: Or maybe a semi-colon
```suggestion
 * in-memory downscaling via `scalePixels`, but discards the full copy of the decoded image. Accounts
```

**Review by Bigpet — APPROVED** [review]
lgtm

File: src/core/UbqScaledImageGenerator.cpp (hunk 5 above)
Comment (maltebaumann) [reply src/core/UbqScaledImageGenerator.cpp:103]: Paranoid isn't always a bad idea 😅

**Pull request merged**